From 62960a89c36155163cb30d086ae4dca3f96fd221 Mon Sep 17 00:00:00 2001
From: snoweay <snoweay@163.com>
Date: Thu, 8 Dec 2022 03:05:32 +0000
Subject: [PATCH] upatch: Disable kmod compile

Signed-off-by: snoweay <snoweay@163.com>
---
 upatch/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/upatch/CMakeLists.txt b/upatch/CMakeLists.txt
index 3e8e127..d063259 100644
--- a/upatch/CMakeLists.txt
+++ b/upatch/CMakeLists.txt
@@ -11,4 +11,4 @@ endif()
 add_subdirectory(upatch-build)
 add_subdirectory(upatch-diff)
 add_subdirectory(upatch-tool)
-add_subdirectory(kmod)
+#add_subdirectory(kmod)
-- 
2.33.0

